vllm/entrypoints/chat_utils.py

```diff
@@ -80,10 +80,61 @@ class ConversationMessage(TypedDict):
     content: str
 
 
-@dataclass(frozen=True)
-class ChatMessageParseResult:
-    messages: List[ConversationMessage]
-    mm_futures: List[Awaitable[MultiModalDataDict]]
+class MultiModalItemTracker:
+    """
+    Tracks multi-model items in a given request and ensures that the number
+    of multi-modal items in a given request does not exceed the configured
+    maximum per prompt.
+    """
+
+    def __init__(self, model_config: ModelConfig):
+        self._allowed_items = (model_config.multimodal_config.limit_per_prompt
+                               if model_config.multimodal_config else {})
+        self._consumed_items = {k: 0 for k in self._allowed_items}
+        self._futures: List[Awaitable[MultiModalDataDict]] = []
+
+    def add(self, modality: Literal["image", "audio"],
+            mm_future: Awaitable[MultiModalDataDict]):
+        allowed_count = self._allowed_items.get(modality, 1)
+        existing_count = self._consumed_items.get(modality, 0)
+        if existing_count >= allowed_count:
+            raise ValueError(
+                f"At most {allowed_count} {modality}s may be provided in one "
+                "request.")
+
+        self._consumed_items[modality] = existing_count + 1
+        self._futures.append(mm_future)
+
+    def all_mm_data(self) -> Optional[Awaitable[MultiModalDataDict]]:
+
+        async def _combine(futures: List[Awaitable[MultiModalDataDict]]):
+            mm_data: MultiModalDataDict = {}
+
+            # Merge all the multi-modal items
+            for single_mm_data in (await asyncio.gather(*futures)):
+                for mm_key, mm_item in single_mm_data.items():
+                    existing_item = mm_data.get(mm_key)
+                    if not existing_item:
+                        # Clone it if it's already a list so we can freely
+                        # mutate it later.
+                        item_to_insert = mm_item[:] if isinstance(
+                            mm_item, list) else mm_item
+                        mm_data[mm_key] = item_to_insert  # type: ignore
+                    else:
+                        if isinstance(existing_item, list):
+                            result_list = existing_item
+                        else:
+                            result_list = [existing_item]
+                            mm_data[mm_key] = result_list  # type: ignore
+
+                        if isinstance(mm_item, list):
+                            result_list.extend(mm_item)
+                        else:
+                            result_list.append(mm_item)
```

Comment:
Looks quite complicated. Can we simplify the code here using a `defaultdict`? We allow `Mapping[str, object]` in `MultiModalDataDict` so this shouldn't cause any type errors.